lib: open-amp: support resource table on C++ applications

dornbirndevelops opened the pull request:
on c++ applications, when using the translation unit resource_table.c it appears that the symbols defined inside cannot be found properly.

stephanosio reviewed (changes requested):
Please add the commit message (body) and sign-off as required by the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines).

dornbirndevelops committed:
lib: open-amp: support resource table on C++ applications

In C++ applications, methods which are meant to be implemented in C
should be wrapped in a `extern "C"` scope at the point of declaration.
This enables the correct symbol table mangling, which fixes current
linker errors in C++.

Signed-off-by: Alexander Dengg <dornbirndevelops@gmail.com>

arnopo reviewed (approved):
LGTM
Thanks!
